magmi/plugins/extra/itemprocessors/imageprocessor/imageitattributeemprocessor.php

```diff
@@ -224,29 +224,38 @@ public function handleVarcharAttribute($pid, &$item, $storeid, $attrcode, $attrd
     }
 
     /**
-     * imageInGallery
+     * Creates or updates image and returns its id.
      *
      * @param int $pid
      *            : product id to test image existence in gallery
      * @param string $imgname
      *            : image file name (relative to /products/media in magento dir)
+     * @param int $refid
+     *            : A reference to an attribute type, typically of image,
+     *            small_image or thumbnail
      * @return bool : if image is already present in gallery for a given product id
      */
     public function getImageId($pid, $attid, $imgname, $refid = null, $store_id = 0)
     {
         $t = $this->tablename('catalog_product_entity_media_gallery');
 
         $sql = "SELECT $t.value_id FROM $t ";
+
+        // Try finding matching refid
         if ($refid != null) {
             $vc = $this->tablename('catalog_product_entity_varchar');
-            $sql .= " JOIN $vc ON $t.entity_id=$vc.entity_id AND $t.value=$vc.value AND $vc.attribute_id=?
+            $full_sql = $sql . " JOIN $vc ON $t.entity_id=$vc.entity_id AND $t.value=$vc.value AND $vc.attribute_id=?
 					WHERE $t.entity_id=? AND $vc.store_id=?";
-            $imgid = $this->selectone($sql, array($refid, $pid, $store_id), 'value_id');
-        } else {
-            $sql .= " WHERE value=? AND entity_id=? AND attribute_id=?";
-            $imgid = $this->selectone($sql, array($imgname, $pid, $attid), 'value_id');
+            $imgid = $this->selectone($full_sql, array($refid, $pid, $store_id), 'value_id');
+        }
+
+        // If not found just match on type and value
+        if ($imgid == null) {
+            $full_sql = $sql . " WHERE value=? AND entity_id=? AND attribute_id=?";
+            $imgid = $this->selectone($full_sql, array($imgname, $pid, $attid), 'value_id');
```

Comment:
So I tried swapping the photo of a product to another photo that already existed on the product in media gallery. This works someone fine. 

The default values on the products are wrong:

* image
* small_image
* thumbnail 

Will all remain as they were before, on the old image. But on the store views these will be correctly swapped to the new image.

(As I see this is a different problem and has nothing to do with what changes I'm proposing here.)

Also magmi doesn't remove the the previous excludes essentially hiding the old small_image and thumbnail from the gallery 👍 .

But the default values should probably be addressed as well in another issue.

In the end if you update an image you end up with a bunch of small_image and thumbnails that are hidden from all stores. Doesn't do any harm but it would be nice to have them go away. 

I've tried with the `media_gallery_reset` to values 0 and 1 and the column completely left out. No difference.